kitsune/sumo/static/sumo/js/switching-devices-wizard-manager.js

```diff
@@ -8,9 +8,113 @@
+      // don't get any passed through to us from the entrypoint to this
+      // SUMO page. These are what will ultimately get passed through the
+      // FxA sign-in flow.
+      utm_source: "support.mozilla.org",
+      utm_campaign: "migration",
+      utm_medium: "mozilla-websites",
+
+      entrypoint: "fx-new-device-sync",
+      entrypoint_experiment: null,
+      entrypoint_variation: null,
+      flow_id: null,
+      flow_begin_time: null,
+      context: null,
+    },
+  };
 
   constructor(formWizard) {
     this.#formWizard = formWizard;
+
+    if (!this.#formWizard.hasAttribute("fxa-root")) {
+      throw new Error(
+        "Expected fxa-root on element passed to " +
+          "SwitchingDevicesWizardManager constructor."
+      );
+    }
+    this.#fxaRoot = this.#formWizard.getAttribute("fxa-root");
+
+    // If the page was loaded with any UTM or entrypoint parameters, we'll want
+    // to pass those along, so now we check for those and add them to our state
+    // if need be. Since this occurs during construction, we don't need to call
+    // #updateState().
+    let searchParams = new URLSearchParams(window.location.search);
+
+    for (let paramName of [
+      "utm_source",
+      "utm_campaign",
+      "utm_medium",
+      "entrypoint",
+      "entrypoint_experiment",
+      "entrypoint_variation",
+    ]) {
+      if (searchParams.has(paramName)) {
+        this.#state.fxa[paramName] = searchParams.get(paramName);
+      }
+    }
+
+    // We need to get some query parameters from the FxA server before we
+    // show the user any kind of form to create or sign-in to an account.
+    // See https://mozilla.github.io/ecosystem-platform/relying-parties/reference/metrics-for-relying-parties#relying-party-hosted-email-form.
+    this.#requestMetricsParams();
+  }
+
+  /**
+   * Returns a fresh copy of the internal state object. This is mainly
+   * for testing.
+   *
+   * @type {object}
+   */
+  get state() {
+    return Object.assign({}, this.#state);
+  }
+
+  /**
+   * Updates the internal state object by replacing it with a copy of newState.
+   *
+   * @param {object}
+   *   The new state object.
+   */
+  #updateState(newState) {
+    this.#state = Object.assign({}, newState);
+  }
+
+  /**
+   * Makes contact with the FxA endpoint pointed at by #fxaRoot/metrics-flow to
+   * get the flow_id and flow_begin_time parameters. Once those parameters are
+   * returned, this triggers a state update.
+   *
+   * @returns {Promise<undefined>}
+   */
+  async #requestMetricsParams() {
+    let params = new URLSearchParams();
+    for (let paramName of [
+      "utm_source",
+      "utm_campaign",
+      "entrypoint",
+      "entrypoint_variation",
+      "entrypoint_experiment",
+    ]) {
+      if (this.#state.fxa[paramName]) {
+        params.set(paramName, this.#state.fxa[paramName]);
+      }
+    }
+    let response = await window.fetch(
+      `${this.#fxaRoot}/metrics-flow?${params}`
```

Comment:
@mikeconley Will `/metrics-flow` be the only path used with `this.#fxaRoot` within `SwitchingDevicesWizardManager`? I'm just wondering if I'll perhaps need to broaden our current [CSP `connect-src`](https://github.com/mozilla/kitsune/blob/d2a822f229a4426d73e42535be39cb816b75b965/kitsune/settings.py#L1221) setting?

Reply:
Via an XHR? Yes, I believe so. I believe we'll end up submitting a `<form>` similar to [this one](https://github.com/mozilla/activity-stream/blob/06aeeb331e9dd497e4d115d0e6cba51b9b25b36c/content-src/asrouter/templates/StartupOverlay/StartupOverlay.jsx#L124C1-L144) to get the user into the sign-in flow, but I don't _believe_ that requires a connect-src CSP change. We may, however, need to add the FxA server to the form-action list: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/form-action

Reply:
I've filed a ticket to get the FxA server added to the form-action CSP list: https://github.com/mozilla/sumo/issues/1264

Reply:
@mikeconley Good catch, thanks!